[mono] Don't call Assembly.CodeBase directly in RuntimeAssembly.GetName

lambdageek opened the pull request:
It's marked as not available in single file apps.
Call the underlying get_code_base icall.

Fixes https://github.com/dotnet/runtime/issues/54835

lambdageek committed:
[mono] Don't call Assembly.CodeBase directly in RuntimeAssembly.GetName

It's marked as not available in single file apps.
Call the underlying get_code_base icall.

Fixes https://github.com/dotnet/runtime/issues/54835

marek-safar commented on src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -252,7 +252,7 @@ private static void AddPublicNestedTypes(Type type, List<Type> types, List<Excep
         public override AssemblyName GetName(bool copiedName)
         {
-            return AssemblyName.Create(_mono_assembly, CodeBase);
+            return AssemblyName.Create(_mono_assembly, get_code_base (this, false));
```
Is this an actual fix? What does it return for single-file like targets?

lambdageek replied:
In Mono we have two names associated with an assembly.  It's "filename" which corresponds to `Assembly.Location` which is NULL for single-file targets like WebAssembly.  And "name" which is derived from the name that the assembly was called in the bundle which is returned for `Assembly.CodeBase`.

So yea this is the actual fix - the issue is that the CodeBase public getter is tagged as not available in single-file.  But `get_code_base` is not.  And it returns a reasonable value.  So we use it directly.

marek-safar replied:
Yeah and I don't think that's correct. The value is exposed via https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyName.cs#L63 which says that the values should be empty string (null value effectively)

lambdageek replied:
Thanks.  Switched to using `MonoImage:filename` which will be null for bundles

lambdageek committed:
[icall] Use MonoImage:filename for RuntimeAssembly.get_code_base

For bundled asssemblies in single file scenarios, RuntimeAssembly.CodeBase will
be null, matching CoreCLR.

lambdageek committed:
Fixup after rebase on origin/main

lambdageek commented:
Ok, so now wasm is failing a codebase test in `src/libraries/System.Reflection/tests/AssemblyTests.cs` because we now return null for the codebase here.

```
        [Fact]
        public void CodeBase()
        {
            Assert.NotEmpty(Helpers.ExecutingAssembly.CodeBase);
        }
```

How does this test pass on CoreCLR single file? or is it skipped somehow?  @eerhardt

eerhardt commented:
> How does this test pass on CoreCLR single file?

I don't believe the test is executed on CoreCLR single file. Honestly, I don't think any of our library tests are executed under CoreCLR single file.

cc @agocke @tlakollo @vitek-karas

tlakollo commented:
> I don't believe the test is executed on CoreCLR single file. Honestly, I don't think any of our library tests are executed under CoreCLR single file.

I believe that is controlled by the following switch -> https://github.com/dotnet/runtime/blob/main/src/libraries/tests.proj#L329
We only tests single file in System.Collections.Tests.csproj and System.IO.IsolatedStorage.Tests.csproj but could be enabled in other projects

marek-safar commented:
> How does this test pass on CoreCLR single file? or is it skipped somehow?

I believe there is a single-file like PlatformDetection property.

> We only tests single file in System.Collections.Tests.csproj and System.IO.IsolatedStorage.Tests.csproj but could be enabled in other projects

I think System.Reflection would be a very useful addition to the list.

lambdageek committed:
disable codebase test on wasm
